File: server/filters/status.c
#include "filters.h"

FILTER_SETUP( status )
{
	char* copy = strdup(args_name);
	char *pt;
	printf("status", args_name);

	pt = strtok(copy, ",");
	while (pt != NULL)
	{
		unsigned char a = strtoul( pt, NULL, 16 );
		callback->args.status.out_status[ a - 0x80 ] = a;
		printf(" 0x%X", a);
		pt = strtok(NULL, ",");
	}
	free(copy);
}
FILTER_CALLBACK( status )
{
	unsigned char *out_status = args->status.out_status;

	int a = 0;
	for( int b = 0; b < n_bytes; ++b )
	{
		unsigned char cur_state = scan_status(data, buf[b]);
		if( out_status[ cur_state - 0x80 ] == cur_state )
		{
			out_buf[a++] = buf[b];
		}
	}
	return write_buffer( data->output_device->midi, out_buf, a );
}

File: rust/src/filters/status.rs
use std::convert::TryInto;

pub struct Status { out_status : [u8 ; 0x80] }
impl Status {
	pub fn new( args : String ) -> Status {
		let mut out_status = [ 0 ; 0x80 ];
		for arg in args.split(",") {
			let status = parse_hex(arg);
			let i : usize = ((0x80 | status) - 0x80).try_into().unwrap();
			out_status[i] = status;
		}
		Status { out_status : out_status }
	}
}
impl CallbackFn for Status {
	fn callback(&self, data : &mut CallbackData, buf : &[u8]) -> usize {
		let mut a = 0;
		let out_status = &self.out_status;

		for c in buf {
			let cur_state = data.output_device.scan_status(*c);
			let i : usize = ((0x80 | cur_state) - 0x80).try_into().unwrap();
			if out_status[i] == cur_state {
				data.buffer[ a ] = *c;
				a += 1;
			}
		}
		data.output_device.send_buffer(&data.buffer[0..a]).unwrap()
	}
}
